### ISSUE TITLE
Question: Any projects based on this that ARE trying to be complete like JSDOM?

### ISSUE THREAD

**Issue by isaacs** [WebReflection/linkedom#198]
This is a cool project!  Context: I'm dealing right now with trying to figure out how to have tests for React components that don't take forever to run, can live in a clean plain old node environment instead of Jest's parallel universe, use tap's API instead of the code poetry like `expect(thing).toHave.something('i can never remember this junk').that.isnt.Null()`, etc.

I have it working now, but the big bottleneck is jsdom.  Every component test takes like _5-10 seconds_ to spin up, at minimum, for a test that runs in 50ms or less.  If they do anything interesting with the dom (which a lot of them do), it gets much worse.  I'm starting to see _why_ jest runs tests in the parallel universe that it does.

It seems like `@testing-library/react` and `@testing-library/react` can _almost_ work, but there are some missing pieces (by design, I realize).  I totally understand not wanting to add those pieces to linkedom, and agree with that principle entirely.

So the question is, is there a project you're aware of that uses linkedom under the hood, but then _does_ add that extra stuff?  Some of what seem to be missing that I've run into:

- Location isn't set.  (Can kinda hack around it by setting `window.location` to a URL object, which is close enough, but a proper Location polyfill would obviously be needed for anything doing actual navigation.)
- Navigator has userAgent, but `navigator.clipboard` is missing.
- Document.createRange() returns an object that doen'st have setStart and setEnd methods.

I'm thinking it'd be possible to create a set of polyfills for all these things, and probably end up with something that would really only lack some of the finer points of JSDOM's tree structure, but still be close enough to simulate a browser for react testing.

Is someone doing this already?  Or, if not, is there some reason you know of why linkedom isn't a suitable base for something like this?

Thanks!

**Comment by WebReflection** [comment]
All your points could be addressed in here ... fancy filing a PR?

**Comment by isaacs** [comment]
I can't promise how soon I'll get to it, but if you're open to PRs in that direction, I'll add it to my todo stack, if no one gets to it sooner :)

Maybe a way to frame this is less "fully and faithfully implement every browser spec" (which is jsdom's intent), but instead "implement enough of the browser spec to work as a seamless drop-in replacement for global-jsdom in `@testing-library/react` and `@testing-library/userevents`".

Still a pretty audacious goal, but could be tackled in parts, and it'd sure be nice to speed up the load times for those testing libraries.

**Comment by WebReflection** [comment]
A mock is a mock … I admire JSDOM ambition and yet I would use a real browser instead if real browsers is what I’m after 🤷 … this project goal is SSR and it’s already a stretch to that goal

**Comment by isaacs** [comment]
Yeah, that's why I ask, I guess. SSR and testing are two different use cases, and while jsdom CAN do both, it does it rather inefficiently, because it's also trying to be a feature complete scraper (which is much more ambitious than either). When I get around to picking this up, I guess we can always split up the layers if it gets too unwieldy.
